File: Caelan.Frameworks.Common/DateTimeRange.fs
namespace Caelan.Frameworks.Common.Classes

open System

type DateTimeRange(start : DateTime, ``end``) = 
    
    /// <summary>
    /// Start date of range
    /// </summary>
    member val Start = (start, ``end``) ||> min with get, set
    
    /// <summary>
    /// End date of range
    /// </summary>
    member val End = (start, ``end``) ||> max with get, set
    
    /// <summary>
    /// Check if requested date is in range
    /// </summary>
    /// <param name="date">The date that needs check</param>
    member public this.IsInRange date = this.Start <= date && this.End >= date
    
    /// <summary>
    /// Check if requested dates are in range
    /// </summary>
    /// <param name="dates">The dates list that needs check</param>
    member public this.IsInRange([<ParamArray>] dates : seq<DateTime>) = dates |> Seq.forall this.IsInRange
    
    /// <summary>
    /// Check if requested range is in range
    /// </summary>
    /// <param name="range">The range that needs check</param>
    member public this.IsInRange(range : DateTimeRange) = this.IsInRange([ range.Start; range.End ])
    
    /// <summary>
    /// Conversion to TimeSpan from Range
    /// </summary>
    member public this.ToTimeSpan = this.End - this.Start

File: Caelan.Frameworks.Common/Classes/DateTimeRange.cs
using System;
using System.Linq;
using System.ServiceModel;

namespace Caelan.Frameworks.Common.Classes
{
    [DataContract]
    public class DateTimeRange
    {
        [DataMember]
        public DateTime Start { get; set; }
        
        [DataMember]
        public DateTime End { get; set; }

        public DateTimeRange(DateTime start, DateTime end)
        {
            if (start > end)
            {
                var tmp = end;
                end = start;
                start = tmp;
            }

            Start = start;
            End = end;
        }

        public bool IsInRange(DateTime date) => Start <= date && End >= date;
        public bool AreInRange(params DateTime[] dates) => dates.All(IsInRange);
        public bool IsInRange(DateTimeRange range) => AreInRange(range.Start, range.End);
        public TimeSpan ToTimeSpan() => End - Start;
    }
}
